lib/refocus/samples.rb

```diff
@@ -5,20 +5,35 @@ module Refocus
   class Samples
     include JsonHelper
 
-    attr_reader :http
+    attr_reader :http, :samples
 
     def initialize(url:, token:)
       @http = Refocus::Http.new(url: url, token: token)
+      @samples = []
     end
 
-    def collector
-      Collector.new(http: http)
+    def upsert(name:, aspect:, value:"", messageBody:"", messageCode:"", relatedLinks:[])
+      sample = format_sample(name:name, aspect: aspect, value: value, messageBody: messageBody, messageCode: messageCode, relatedLinks: relatedLinks)
+      json(http.post("upsert", body: sample, expects: 200))
     end
 
-    def submit(name:, aspect:, value:)
-      c = collector
-      c.add(name: name, aspect: aspect, value: value)
-      c.submit
+    def add_to_bulk(name:, aspect:, value:"", messageBody:"", messageCode:"", relatedLinks:[])
+      samples << format_sample(name:name, aspect: aspect, value: value, messageBody: messageBody, messageCode: messageCode, relatedLinks: relatedLinks)
+    end
+
+    def upsert_bulk
+      if samples.empty?
+        raise Exception, "No samples to upsert have been added. Please use the add_to_bulk method to add samples to upsert first."
+      else
+        result = json(http.post("upsert/bulk", body: samples, expects: 200))
+        samples.clear
+        result
+      end
+    end
```

Comment:
@kkim-sf First of all: Thanks for helping out and welcome! :)

I'd like to elaborate a bit on the pattern I'm promoting here. It looks like in order to use what you've done effectively, you'll need to do something like this:

```ruby
  refocus.samples.add_to_bulk(name:"mysample", aspect: "myaspect", value: 1)
  refocus.samples.add_to_bulk(name:"mysample", aspect: "myaspect", value: 2)
  refocus.samples.upsert_bulk
```

The problem I have with this is that this is that the collector already enables this sort of semantics - ie, "Add samples one at a time, then submit all at once", like this:

```ruby
   collector = refocus.samples.collector
   collector.add(name: "mysample", aspect: "myaspect")
  collector.submit
```

I'd rather not have the client itself be responsible for maintaining any state other than knowledge of the endpoint it is talking to. The collector should be used to store state like that, which is what it does.

I like the addition of the new parameters - I just think the should be added to the existing `sample.submit` method, and passed through to `Collector.submit`. Submission of a single sample could also be done using its own method in the Samples class instead of passing the values through to a one-off collector object, but I don't think it really matters much for efficiency (ultimately, it would hit the same API). I implemented it this way to avoid rewriting the same code to submit samples twice.

Btw, feel free to reach out on gchat or slack if you want more synchronous communication! I'm in Australia though, so we might have trouble with timezones.

Also - value is optional? Good to know, but it strikes me as strange. What does it even mean to have a sample without a value? 

If you prefer `upsert` instead of `submit` for the method name, I propose maintaining compatibility by having both, perhaps with `alias_method`.